sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ZOrder.scala

```diff
@@ -0,0 +1,222 @@
+   * TODO(SPARK-37363): Support [[StringType]] for Z-order.
+   */
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.size)(IntegralType)
+
+  override def dataType: DataType = BinaryType
+
+  override def prettyName: String = "zorder"
+
+  /**
+   * Number of bits per input in binary representation.
+   */
+  @transient private lazy val inputBitSizes = children.map(_.dataType.defaultSize * 8)
+
+  /**
+   * Maximal number of bits per input in binary representation.
+   */
+  @transient private lazy val inputBitMaxSize = inputBitSizes.max
+
+  /**
+   * Total number of bits per input in binary representation.
+   */
+  @transient private lazy val inputBitTotalSize = inputBitSizes.sum
+
+  /**
+   * The output for Z-order, represented in byte array format.
+   */
+  @transient private lazy val outputBytes = new Array[Byte](inputBitTotalSize / 8)
+
+  /**
+   * Handle input with null value. Convert null into minimal value per each data type.
+   * This mimics [[NullsFirst]] sort ordering.
+   */
+  @transient private lazy val handleNullFromInputs: Seq[Any => Any] = children.map {
+    child => child.dataType match {
+      case ByteType =>
+        value: Any => if (value == null) Byte.MinValue else value
+      case ShortType =>
+        value: Any => if (value == null) Short.MinValue else value
+      case IntegerType =>
+        value: Any => if (value == null) Int.MinValue else value
+      case LongType =>
+        value: Any => if (value == null) Long.MinValue else value
+      case x =>
+        throw new IllegalArgumentException(
+          s"ZOrder expression should not take $x as the data type")
+    }
+  }
+
+  /**
+   * Get the bit in `ordinal` position (0-based) of input `value`.
+   * The return value of corresponding bit is in [[Byte]] type.
+   */
+  @transient private lazy val getBitFromInputs: Seq[(Any, Int) => Byte] = children.map { child =>
+    child.dataType match {
+      case ByteType =>
+        (value: Any, ordinal: Int) =>
+          ((value.asInstanceOf[Byte] >> ordinal) & 1).toByte
+      case ShortType =>
+        (value: Any, ordinal: Int) =>
+          ((value.asInstanceOf[Short] >> ordinal) & 1).toByte
+      case IntegerType =>
+        (value: Any, ordinal: Int) =>
+          ((value.asInstanceOf[Int] >> ordinal) & 1).toByte
+      case LongType =>
+        (value: Any, ordinal: Int) =>
+          ((value.asInstanceOf[Long] >> ordinal) & 1).toByte
+      case x =>
+        throw new IllegalArgumentException(
+          s"ZOrder expression should not take $x as the data type")
+    }
+  }
+
+  /**
+   * Get the Z-order sign bit of input `value`.
+   *
+   * For [[IntegralType]], Z-order flips the original sign bit of input. Sign bit is 0 for negative
+   * value, and 1 for positive value. This is to maintain the same ordering between input and
+   * output in [[BinaryType]].
+   */
+  @transient private lazy val getZOrderSignBitFromInputs: Seq[Any => Byte] = children.map {
```

Comment:
this will involve a lot of object creation and gc-ed. 
is it possible to put all the input bytes in one byte[] and do interleaving by accessing the input byte[] directly?

Reply:
> this will involve a lot of object creation and gc-ed.

Please correct me if my understanding is wrong, but here with `lazy val`, `getZOrderSignBitFromInputs` is only created once and reused per row.

Reply:
```
case ByteType =>
        value: Any => if (value.asInstanceOf[Byte] < 0) 0.toByte else 1.toByte
```
the value is an Any type, which will involves boxing and unboxing, which involves object creation. For code generation optimizations, we would avoid this type of methods.